barretenberg/cpp/src/barretenberg/vm/avm/generated/circuit_builder.cpp

```diff
@@ -30,15 +50,37 @@ AvmCircuitBuilder::ProverPolynomials AvmCircuitBuilder::compute_polynomials() co
                                               /*make shiftable with offset*/ 1 };
                        }
                    }));
+
     // catch-all with fully formed polynomials
     AVM_TRACK_TIME(
         "circuit_builder/init_polys_unshifted", ({
             auto unshifted = polys.get_unshifted();
+
+            // Set of the labels for derived/inverse polynomials.
+            const auto derived_labels = polys.get_derived_labels();
+            std::set<std::string> derived_labels_set(derived_labels.begin(), derived_labels.end());
+
             bb::parallel_for(unshifted.size(), [&](size_t i) {
                 auto& poly = unshifted[i];
+                const auto col_idx = polys_to_cols_unshifted_idx[i];
+                size_t col_size = 0;
+
+                // We fully allocate the inverse polynomials. We leave this potential memory optimization for later.
+                if (derived_labels_set.contains(labels[i])) {
+                    col_size = num_rows;
+                } else {
+                    // We process each column from bottom and stops as soon as we encounter a non-zero entry.
+                    for (size_t j = 0; j < num_rows; j++) {
+                        if (!rows[num_rows - j - 1].as_vector()[col_idx].is_zero()) {
```

Comment:
as_vector is expensive, a pity that it cannot be done without it.

Reply:
Yep I did not see another way if we want to parallelize over the columns.